### PR TITLE
fix an infinite loop which happens when autometrics is initialized in a node service without a git repository

### PR DESCRIPTION
**Pull request by benjibuiltit**
after instrumenting my node / ts server with autometrics, things seemed to be working just fine locally.

unfortunately when i deployed to my dev environment (kubernetes cluster) i saw my service spike in cpu utilization and the event loop would hang causing all requests to time out.

this made it difficult to debug since i was unable to reproduce locally initially. i found that i could reproduce this behavior if i built the docker image and ran the container locally. from there i was able to run node with profiling flags to get the v8 log output. this pointed me to the `readFileSync` call inside the `readClosest` function and the issue became clear. there is an unhandled case on the node platform that will result in an infinite while loop if no git repository is initialized (as is the case in a container).

this fixes that, but i'm uncertain if the other differences found between the [node](https://github.com/autometrics-dev/autometrics-ts/blob/06ff7739ccd82f2a1d382f3b0d62c906ea72956a/packages/autometrics/src/platform.node.ts#L136-L144) and [deno](https://github.com/autometrics-dev/autometrics-ts/blob/06ff7739ccd82f2a1d382f3b0d62c906ea72956a/packages/autometrics/src/platform.deno.ts#L130-L144) platforms should be ported as well.

additionally, i had a troublesome time getting the `just build-all` command to work when trying to prepare this for a proper release/pr. please let me know how to proceed with getting this merged in. it is currently blocking my adoption of autometrics which otherwise looks to be of great benefit to me.

node profile ouput (sample)
```
 [Bottom up (heavy) profile]:
  Note: percentage shows a share of a particular caller in the total
  amount of its parent calls.
  Callers occupying less than 1.0% are not shown.

   ticks parent  name
   9335   64.6%  /usr/local/bin/node
   4380   46.9%    LazyCompile: *uvException node:internal/errors:511:57
   4380  100.0%      LazyCompile: *readFileSync node:fs:468:22
   4371   99.8%        LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
   4371  100.0%          LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
   4371  100.0%            LazyCompile: ~getRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:517:30
   3742   40.1%    LazyCompile: *readFileSync node:fs:468:22
   3735   99.8%      LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
   3735  100.0%        LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
   3735  100.0%          LazyCompile: ~getRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:517:30
   3735  100.0%            LazyCompile: ~getDefaultValue /app/node_modules/@autometrics/autometrics/index.cjs:715:25
    299    3.2%    /usr/local/bin/node
    270   90.3%      LazyCompile: *readFileSync node:fs:468:22
    270  100.0%        LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
    270  100.0%          LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
    270  100.0%            LazyCompile: ~getRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:517:30
     10    3.3%      LazyCompile: *uvException node:internal/errors:511:57
     10  100.0%        LazyCompile: *readFileSync node:fs:468:22
     10  100.0%          LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
     10  100.0%            LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
    280    3.0%    LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
    280  100.0%      LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
    280  100.0%        LazyCompile: ~getRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:517:30
    280  100.0%          LazyCompile: ~getDefaultValue /app/node_modules/@autometrics/autometrics/index.cjs:715:25
    280  100.0%            LazyCompile: ~recordBuildInfo /app/node_modules/@autometrics/autometrics/index.cjs:695:29
    170    1.8%    LazyCompile: *isErrorStackTraceLimitWritable node:internal/errors:205:40
    170  100.0%      LazyCompile: *uvException node:internal/errors:511:57
    170  100.0%        LazyCompile: *readFileSync node:fs:468:22
    167   98.2%          LazyCompile: *readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
    167  100.0%            LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
      3    1.8%          Function: ^readClosest /app/node_modules/@autometrics/autometrics/index.cjs:566:21
      3  100.0%            LazyCompile: ~detectRepositoryUrl /app/node_modules/@autometrics/autometrics/index.cjs:560:29
    105    1.1%    Function: ^internalCompileFunction node:internal/vm:31:33
    104   99.0%      Function: ^wrapSafe node:internal/modules/cjs/loader:1154:18
    104  100.0%        Function: ^Module._compile node:internal/modules/cjs/loader:1210:37
    103   99.0%          Function: ^Module._extensions..js node:internal/modules/cjs/loader:1265:37
    102   99.0%            Function: ^Module.load node:internal/modules/cjs/loader:1107:33
```

`just build-all` error
```
~/Development/autometrics-ts/examples/fastify ~/Development/autometrics-ts
+ just build
yarn && yarn build
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: ENOENT: no such file or directory, readlink '/Users/benji/Development/autometrics-ts/dist/autometrics/node_modules/@types/node'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'readlink',
  path: '/Users/benji/Development/autometrics-ts/dist/autometrics/node_modules/@types/node'
}
error: Recipe `build` failed on line 2 with exit code 1
error: Recipe `build-examples` failed with exit code 1
```

### CODE REVIEW

**Comment by brettimus** [comment]
really good catch! this must've been a pain to debug. i think the solution needs a slight tweak, will comment on the file itself, then look into the `build-all` command error

File: packages/autometrics/src/platform.node.ts [hunk 1]
```diff
@@ -140,6 +140,7 @@ function readClosest(path: string): Uint8Array {
       return readFileSync(join(basePath, path));
     } catch {
       basePath = dirname(basePath);
+      break;
```
Comment (brettimus) [inline packages/autometrics/src/platform.node.ts:143]: so i think we need to modify the `while` condition instead of adding a break here, since the break would prevent us from recursively searching parent directories. 

the infinite loop you encountered seems to happen because `dirname("/") === "/"`, so we'll never reach the breaking case of `basePath.length === 0` in the while condition when we're dealing with full paths

so, we need a base case in the while loop that covers when the basePath is reduced to `"/"`. but also to be fully cross-platform, we should consider what the equivalent windows file path(s) might look like

Comment (brettimus) [reply packages/autometrics/src/platform.node.ts:143]: also worth noting is the node.js `dirname` function returns `"."` for the empty string (https://github.com/nodejs/node/blob/0161ad0baf87a0009101ce00b22b874ad6fc5d88/lib/path.js#L660)

and there's some special handling in `dirname` for windows device roots (see [here](https://github.com/nodejs/node/blob/main/lib/path.js#L718))

i might say for now we create a util function `isRootPath` that can also be used in the while loop's condition. 

basically, we need to beak the loop when `dirname` hits a root

Comment (brettimus) [reply packages/autometrics/src/platform.node.ts]: perfect — that's a good cross-platform solution. and indeed it was better to keep that break inside the catch (instead of the while loop condition), to ensure we also actually check the fs root

**Review by brettimus — APPROVED** [review]

File: packages/autometrics/src/platform.node.ts [hunk 2]
```diff
@@ -112,6 +112,15 @@ export function getALSInstance() {
   }>();
 }
 
+/**
+ * Returns a boolean indicating whether a given path is the file-system root or not.
+ *
+ * @internal
+ */
+export function isRootPath(pathToCheck: str): string {
```
Comment (brettimus) [inline packages/autometrics/src/platform.node.ts:120]: oh i didn't actually pull down the code, but the type here should be `:string` instead of `:str`

**Comment by brettimus** [comment]
approved but if you could fix the type annotation in the new util function that'd be great!

File: packages/autometrics/src/platform.node.ts (hunk 2 above)
Comment (benjibuiltit) [reply packages/autometrics/src/platform.node.ts:120]: i've been writing a lot of python lately 😁

**Review by keturiosakys — APPROVED** [review]

**Comment by brettimus** [comment]
alright i did get the `just` script to work locally, kinda hackily. what i did was:

- delete `node_modules` (because i had a crufty copy)
- update `deno` locally
- bust all cached deno files with a bash one-liner `find . -type f \( -name "*.ts" -o -name "*.js" \) -exec deno cache --reload {} +`

after that, running `just` worked for me

**Comment by brettimus** [comment]
the plan for now is that we'll merge your PRs, then i'm going to open a separate one to prep a release. we should have these changes released later this week!
